module/zfs/zil.c

```diff
@@ -783,6 +783,9 @@ zil_destroy(zilog_t *zilog, boolean_t keep_first)
 	mutex_exit(&zilog->zl_lock);
 
 	dmu_tx_commit(tx);
+
+	if (zilog->zl_os->os_encrypted)
+		txg_wait_synced(zilog->zl_dmu_pool, txg);
```

Comment:
I think it's worth a comment here to explain:

1. why we need to do this for encrypted datasets
2. what could happen if we didn't do this
3. why we don't do this for non-encrypted datasets (e.g. performance? correctness? something else?)